File: mesh_service_example/src/main/java/com/meshtastic/android/meshserviceexample/ble/HrPlxBleClient.java
package com.meshtastic.android.meshserviceexample.ble;

import android.annotation.SuppressLint;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothGatt;
import android.bluetooth.BluetoothGattCallback;
import android.bluetooth.BluetoothGattCharacteristic;
import android.bluetooth.BluetoothGattDescriptor;
import android.bluetooth.BluetoothGattService;
import android.bluetooth.BluetoothProfile;
import android.content.Context;
import android.util.Log;

import java.util.UUID;

/**
 * Manages connection and data transfer with a BLE Heart Rate and Pulse Oximeter device.
 */
@SuppressLint("MissingPermission")
public class HrPlxBleClient {
    private static final String TAG = "HrPlxBleClient";

    private final Context context;
    private BluetoothGatt bluetoothGatt;
    private Listener listener;

    public interface Listener {
        void onConnectionStateChanged(int state);
        void onServicesDiscovered();
        void onCharacteristicRead(String uuid, byte[] value);
        void onCharacteristicChanged(String uuid, byte[] value);
        void onHeartRateReceived(int hr);
        void onSpo2Received(Integer spo2); // Must be implemented as per error log
    }

    public HrPlxBleClient(Context context) {
        this.context = context;
    }

    public void connect(BluetoothDevice device, Listener listener) {
        this.listener = listener;
        device.connectGatt(context, false, gattCallback);
    }

    public void disconnect() {
        if (bluetoothGatt != null) {
            bluetoothGatt.disconnect();
        }
    }

    private final BluetoothGattCallback gattCallback = new BluetoothGattCallback() {
        @Override
        public void onConnectionStateChange(BluetoothGatt gatt, int status, int newState) {
            if (listener != null) {
                listener.onConnectionStateChanged(newState);
            }
            if (newState == BluetoothProfile.STATE_CONNECTED) {
                Log.i(TAG, "Connected to GATT server.");
                bluetoothGatt = gatt;
                gatt.discoverServices();
            } else if (newState == BluetoothProfile.STATE_DISCONNECTED) {
                Log.i(TAG, "Disconnected from GATT server.");
                if (bluetoothGatt != null) {
                    bluetoothGatt.close();
                    bluetoothGatt = null;
                }
            }
        }

        @Override
        public void onServicesDiscovered(BluetoothGatt gatt, int status) {
            if (status == BluetoothGatt.GATT_SUCCESS) {
                Log.i(TAG, "Services discovered.");
                if (listener != null) {
                    listener.onServicesDiscovered();
                }
                enableNotifications(gatt, BleUtils.HEART_RATE_SERVICE_UUID, BleUtils.HEART_RATE_MEASUREMENT_CHAR_UUID);
            } else {
                Log.w(TAG, "onServicesDiscovered received: " + status);
            }
        }

        @Override
        public void onCharacteristicChanged(BluetoothGatt gatt, BluetoothGattCharacteristic characteristic) {
            if (listener != null) {
                listener.onCharacteristicChanged(characteristic.getUuid().toString(), characteristic.getValue());
            }

            if (BleUtils.HEART_RATE_MEASUREMENT_CHAR_UUID.equals(characteristic.getUuid())) {
                parseHeartRate(characteristic.getValue());
            }
        }
    };

    private void enableNotifications(BluetoothGatt gatt, UUID serviceUuid, UUID characteristicUuid) {
        BluetoothGattService service = gatt.getService(serviceUuid);
        if (service == null) {
            Log.e(TAG, "Service not found: " + serviceUuid);
            return;
        }
        BluetoothGattCharacteristic characteristic = service.getCharacteristic(characteristicUuid);
        if (characteristic == null) {
            Log.e(TAG, "Characteristic not found: " + characteristicUuid);
            return;
        }

        gatt.setCharacteristicNotification(characteristic, true);
        BluetoothGattDescriptor descriptor = characteristic.getDescriptor(BleUtils.CLIENT_CHARACTERISTIC_CONFIG_UUID);
        descriptor.setValue(BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE);
        gatt.writeDescriptor(descriptor);
        Log.d(TAG, "Enabled notifications for " + characteristicUuid);
    }

    private void parseHeartRate(byte[] value) {
        if (value != null && value.length > 1) {
            int flag = value[0];
            int format = (flag & 0x01) == 1 ? BluetoothGattCharacteristic.FORMAT_UINT16 : BluetoothGattCharacteristic.FORMAT_UINT8;
            int hr = (format == BluetoothGattCharacteristic.FORMAT_UINT16) ? (value[1] & 0xFF) | ((value[2] & 0xFF) << 8) : value[1] & 0xFF;

            if (listener != null) {
                listener.onHeartRateReceived(hr);
            }
        }
    }
}

File: mesh_service_example/src/main/java/com/meshtastic/android/meshserviceexample/ble/HrPlxBleClient.kt
package com.meshtastic.android.meshserviceexample.ble

import android.bluetooth.*
import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.Log
import java.util.UUID
import kotlin.math.min

class HrPlxBleClient(private val context: Context) {

    interface Listener {
        fun onConnectionChanged(connected: Boolean)
        fun onHr(hr: Int)
        fun onSpo2(spo2: Int?)
    }

    companion object {
        val UUID_HR_SERVICE: UUID = UUID.fromString("0000180D-0000-1000-8000-00805F9B34FB")
        val UUID_HR_MEAS:   UUID = UUID.fromString("00002A37-0000-1000-8000-00805F9B34FB")

        // 可選：血氧（連續量測 0x2A5F）
        val UUID_PLX_SERVICE: UUID    = UUID.fromString("00001822-0000-1000-8000-00805F9B34FB")
        val UUID_PLX_CONTINUOUS: UUID = UUID.fromString("00002A5F-0000-1000-8000-00805F9B34FB")

        private val UUID_CCCD: UUID = UUID.fromString("00002902-0000-1000-8000-00805F9B34FB")

        private const val TAG = "HrPlxBleClient"
    }

    private val main = Handler(Looper.getMainLooper())

    @Volatile private var listener: Listener? = null
    fun setListener(l: Listener?) { listener = l }

    var connectedAddress: String? = null
        private set

    private var gatt: BluetoothGatt? = null

    /** 連線：Java -> bleClient.connect(device, new Listener{...}); */
    fun connect(device: BluetoothDevice, l: Listener): Boolean {
        listener = l
        // 清理舊連線
        try { gatt?.disconnect() } catch (_: Exception) {}
        try { gatt?.close() } catch (_: Exception) {}
        gatt = null
        connectedAddress = device.address

        gatt = device.connectGatt(context, /*autoConnect*/ false, gattCallback)
        return gatt != null
    }

    fun disconnect() {
        try {
            gatt?.services?.forEach { srv ->
                srv.characteristics?.forEach { ch ->
                    ch.getDescriptor(UUID_CCCD)?.let { d ->
                        try {
                            gatt?.setCharacteristicNotification(ch, false)
                            d.value = BluetoothGattDescriptor.DISABLE_NOTIFICATION_VALUE
                            gatt?.writeDescriptor(d)
                        } catch (_: Exception) {}
                    }
                }
            }
        } catch (_: Exception) {}
        try { gatt?.disconnect() } catch (_: Exception) {}
        try { gatt?.close() } catch (_: Exception) {}
        gatt = null
        if (connectedAddress != null) {
            connectedAddress = null
            post { listener?.onConnectionChanged(false) }
        }
    }

    private val gattCallback = object : BluetoothGattCallback() {
        override fun onConnectionStateChange(g: BluetoothGatt, status: Int, newState: Int) {
            if (newState == BluetoothProfile.STATE_CONNECTED) {
                gatt = g
                post { listener?.onConnectionChanged(true) }
                // 某些機型需要延遲後再 discover
                main.postDelayed({ g.discoverServices() }, 250)
            } else if (newState == BluetoothProfile.STATE_DISCONNECTED) {
                post { listener?.onConnectionChanged(false) }
                try { g.close() } catch (_: Exception) {}
                gatt = null
                connectedAddress = null
            }
        }

        override fun onServicesDiscovered(g: BluetoothGatt, status: Int) {
            enableNotify(g, UUID_HR_SERVICE, UUID_HR_MEAS)
            // 如需 SpO2 再打開
            enableNotify(g, UUID_PLX_SERVICE, UUID_PLX_CONTINUOUS)
        }

        override fun onCharacteristicChanged(
            g: BluetoothGatt, ch: BluetoothGattCharacteristic
        ) {
            when (ch.uuid) {
                UUID_HR_MEAS -> parseHr(ch.value)?.let { bpm ->
                    post { listener?.onHr(bpm) }
                }
                UUID_PLX_CONTINUOUS -> {
                    val (spo2, pr) = parsePlx(ch.value)
                    post { listener?.onSpo2(spo2) }
                    // 如需要 PR 可擴充 callback
                }
            }
        }
    }

    private fun enableNotify(g: BluetoothGatt, srvUuid: UUID, chrUuid: UUID) {
        val c = g.getService(srvUuid)?.getCharacteristic(chrUuid) ?: return
        try {
            g.setCharacteristicNotification(c, true)
            val d = c.getDescriptor(UUID_CCCD) ?: return
            d.value = BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE
            g.writeDescriptor(d)
        } catch (e: Exception) {
            Log.w(TAG, "enableNotify failed: ${e.message}")
        }
    }

    /** Heart Rate Measurement (0x2A37) 解析 */
    private fun parseHr(data: ByteArray?): Int? {
        if (data == null || data.isEmpty()) return null
        val flags = data[0].toInt() and 0xFF
        val sixteenBit = (flags and 0x01) != 0
        return if (!sixteenBit) {
            if (data.size >= 2) data[1].toInt() and 0xFF else null
        } else {
            if (data.size >= 3) {
                ((data[2].toInt() and 0xFF) shl 8) or (data[1].toInt() and 0xFF)
            } else null
        }
    }

    /** 簡化版：PLX Continuous Measurement (0x2A5F) 只抓 SpO2/PR（若有）*/
    private fun parsePlx(data: ByteArray?): Pair<Int?, Int?> {
        if (data == null || data.size < 2) return null to null
        // 這個特徵的 flags/欄位較複雜，這裡示意性抓最常見順序（SpO2 1byte, PR 1byte）
        val spo2 = data[0].toInt() and 0xFF
        val pr   = data[1].toInt() and 0xFF
        return spo2 to pr
    }

    private fun post(block: () -> Unit) = main.post(block)
}
